### ISSUE TITLE
Prototype kcp apiserver -> pcluster egress

### ISSUE THREAD

**Issue by sttts** [kcp-dev/kcp#333]
The kcp apiserver has to talk to the pcluster for
- `kubectl exec|attach|logs`
- webhooks in CRD conversion
- webhooks in admission.

For that kubernetes uses egress, compare https://kubernetes.io/docs/tasks/extend-kubernetes/setup-konnectivity/.

The task here is to enable kcp to make use of this mechanism, configured for each shard (not customizable for the users).

### Acceptance criteria:

1. `kubectl exec/attach/logs` works
2. webhooks work

**Comment by jmprusi** [comment]
I will give this a try!

**Comment by ncdc** [comment]
@pweil- @sttts do you recall if we put this on the roadmap, and if so, which prototype?

**Comment by pweil-** [comment]
At least the logs portion of this has been discussed as priority for prototype 4 as part of the need for basic live observability with no history.    Might consider if we need the full package or can implement items in the description incrementally.

**Comment by ncdc** [comment]
Thanks, added to 4 for now.

**Comment by guymguym** [comment]
From my experience with WebRTC (few years ago) it is very complex (need to set up stun/turn servers to really bypass NAT, and it wasn't ready for enterprise grade systems. Throughput was quite low, which was OK for p2p where each peer was over "slow" internet connections, but definitely not for datacenter connectivity.

Maybe the syncer can be configured to open a persistent connection to KCP, which will allow KCP to make on-demand requests to the cluster api server? Similar to how websockets work, just using grpc...

**Comment by ncdc** [comment]
> Maybe the syncer can be configured to open a persistent connection to KCP, which will allow KCP to make on-demand requests to the cluster api server? Similar to how websockets work, just using grpc...

That's my initial thinking as well

**Comment by aojea** [comment]
> From my experience with WebRTC (few years ago) it is very complex (need to set up stun/turn servers to really bypass NAT in different network levels), and it wasn't ready for enterprise grade systems. Throughput was quite low, which was OK for p2p where each peer was over "slow" internet connections, but definitely not for datacenter connectivity.

Yeah, it is too complex, I've managed to have it working https://github.com/aojea/webrtc-p2p but it requires a custom signaling protocol and integrating it in a `kubectl exec/attach/get`  workflow has a bad UX, the time to setup a connection is considerable.

> and it wasn't ready for enterprise grade systems.

**Comment by sttts** [comment]
Slipping into v0.7.

**Comment by sttts** [comment]
A drawing showing the plumbing into the handler chain: https://docs.google.com/drawings/d/1KCqpg5XJPa7QTJPBKY2RO2Na6HGVspJtVepyqjBsu5o/edit

**Comment by aojea** [comment]
I have the implementation working https://github.com/kcp-dev/kcp/pull/873, however, I can not find the right place to plumb it in the handlers chain.

If someone can pair program with me for 1 or 2 hours we can have it in,

**Comment by ncdc** [comment]
Please set up a time with me today or tomorrow if that works for you

**Comment by ncdc** [comment]
@aojea could you please put this behind a feature gate this is off by default?

**Comment by aojea** [comment]
> @aojea could you please put this behind a feature gate this is off by default?

yes
